### PR TITLE
[REEF-309] Add deadlock information to thread dump on driver force close

### PR DESCRIPTION
**Pull request by dafrista**
This PR adds deadlock information by adding a new static method to ThreadLogger.
The information will only be logged when a deadlock exists.

JIRA:
  [REEF-309](https://issues.apache.org/jira/browse/REEF-309)

Sample actual output of the new output from ThreadLogger after a deadlock:
```
May 07, 2015 8:08:59 AM org.apache.reef.runtime.common.driver.defaults.DefaultClientCloseHandler onNext
WARNING: Deadlocked threads:
        Thread 'Node-1-1430986082981-pool-13-thread-1' with state BLOCKED
                at org.apache.reef.wake.time.runtime.RuntimeClock.stop(RuntimeClock.java:103)
                - waiting to lock: java.util.TreeSet@25a732a4 held by main
                at org.apache.reef.runtime.common.driver.DriverStatusManager.onError(DriverStatusManager.java:146)
                - locked: org.apache.reef.runtime.common.driver.DriverStatusManager@23dc652
                at org.apache.reef.runtime.common.driver.DriverExceptionHandler.onNext(DriverExceptionHandler.java:51)
                at org.apache.reef.runtime.common.driver.DriverExceptionHandler.onNext(DriverExceptionHandler.java:33)
                at org.apache.reef.util.ExceptionHandlingEventHandler.onNext(ExceptionHandlingEventHandler.java:47)
                at org.apache.reef.runtime.common.utils.DispatchingEStage$1.onNext(DispatchingEStage.java:67)
                at org.apache.reef.runtime.common.utils.DispatchingEStage$1.onNext(DispatchingEStage.java:64)
                at org.apache.reef.wake.impl.ThreadPoolStage$1.run(ThreadPoolStage.java:180)
                at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
                at java.util.concurrent.FutureTask.run(FutureTask.java:262)
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
                at java.lang.Thread.run(Thread.java:745)
                * holds locked synchronizer: java.util.concurrent.ThreadPoolExecutor$Worker@38797817
        Thread 'main' with state BLOCKED
                at org.apache.reef.runtime.common.driver.idle.DriverIdleManager.onPotentiallyIdle(DriverIdleManager.java:49)
                - waiting to lock: org.apache.reef.runtime.common.driver.DriverStatusManager@23dc652 held by Node-1-1430986082981-pool-13-thread-1
                - locked: org.apache.reef.runtime.common.driver.idle.DriverIdleManager@3edc04c
                at org.apache.reef.runtime.common.driver.idle.ClockIdlenessSource.onNext(ClockIdlenessSource.java:58)
                - locked: org.apache.reef.runtime.common.driver.idle.ClockIdlenessSource@5ca39971
                at org.apache.reef.runtime.common.driver.idle.ClockIdlenessSource.onNext(ClockIdlenessSource.java:31)
                at org.apache.reef.wake.impl.PubSubEventHandler.onNext(PubSubEventHandler.java:98)
                at org.apache.reef.wake.time.runtime.RuntimeClock.run(RuntimeClock.java:199)
                - locked: java.util.TreeSet@25a732a4
                at org.apache.reef.runtime.common.launch.LaunchClass.run(LaunchClass.java:161)
                at org.apache.reef.runtime.common.Launcher.main(Launcher.java:111)
```

### CODE REVIEW

**Comment by dafrista** [comment]
I'll rebuild this on our windows Jenkins a few times to see if we can get a reef-tests failures. I'd like to see the relevant logs appear on the CI server.

**Comment by dafrista** [comment]
I haven't been able to reproduce the Windows CI server failure I was hoping to. It looks like another error also appears on the CI servers. But hopefully this PR will be useful when we do observe it.

To recap, I'm ready for a review and merge.

File: lang/java/reef-common/src/main/java/org/apache/reef/util/DeadlockInfo.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Provides a view into deadlocked threads for logging or debugging purposes.
+ * Backed by ThreadMXBean
+ */
+public final class DeadlockInfo {
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/util/DeadlockInfo.java]: Make this class package-private. It doesn't seem to be used elsewhere.

File: lang/java/reef-common/src/test/java/org/apache/reef/util/DeadlockInfoTest.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+/**
+ * Test DeadlockInfo
+ */
+public final class DeadlockInfoTest {
+  private static final Logger LOG = Logger.getLogger(DeadlockInfoTest.class.getName());
+
+  private final long millisBeforeDeadlock = 100;
```
Comment (markusweimer) [inline lang/java/reef-common/src/test/java/org/apache/reef/util/DeadlockInfoTest.java]: Can we speed these up? As we add more and more of these real time units based tests, we add to the testing time.

**Comment by markusweimer** [comment]
@dafrista I did a pass. Mostly cosmetics.

**Comment by dafrista** [comment]
Thanks @markusweimer, I've pushed a commit based on your comments.

**Comment by dafrista** [comment]
@markusweimer I settled on 50ms for the timeout. 10ms caused the windows build to fail.

**Comment by markusweimer** [comment]
+1 from me on merging this. However, I can't do the merge in the next few days as I am moving. Feel free to self-merge this one or have another comitter do it.

**Comment by bgchun** [comment]
@markusweimer @dafrista Let me merge it.

**Pull request closed without merging**